docs/rfcs/0012-introduce-ui-store.md

```diff
@@ -0,0 +1,177 @@
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// store thread
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// store thread
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // perform side-effects that also happen on the store thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+    // perform work on the main thread.
+  }
+}
+```
+
+With the changes in this RFC, this switching of threads can be reduced (notable comments marked with 📝):
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // 📝 perform side-effects that now happen on the main thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+Additionally, from [performance investigations already done][2], we know that Fenix creates over a hundred threads within a few seconds of startup. Reducing the number of threads for Stores that do not have a strong requirement to run on a separate thread will lower the applications memory footprint.
+
+## Guide-level explanation
+
+Extending the existing `Store` class to use the `Dispatchers.Main.immediate` will ensure that UI stores will stay on the same UI thread and have that work done immediately. Using a distinct class named `UiStore` also makes it clear to the developer that this is work that will be done on the UI thread and its implications will be made a bit more clear when it's used.
+
+```kotlin
+@MainThread
+open class UiStore<S : State, A : Action>(
+  initialState: S,
+  reducer: Reducer<S, A>,
+  middleware: List<Middleware<S, A>> = emptyList(),
+) : Store<S, A>(
+  initialState,
+  reducer,
+  middleware,
+  UiStoreDispatcher(),
+)
+
+open class Store<S : State, A : Action> internal constructor(
+  initialState: S,
+  reducer: Reducer<S, A>,
+  middleware: List<Middleware<S, A>>,
+  dispatcher: StoreDispatcher,
+) {
+  constructor(
+    initialState: S,
+    reducer: Reducer<S, A>,
+    middleware: List<Middleware<S, A>> = emptyList(),
+    threadNamePrefix: String? = null,
+  ) : this(
+    initialState = initialState,
+    reducer = reducer,
+    middleware = middleware,
+    dispatcher = DefaultStoreDispatcher(threadNamePrefix),
+  )
+}
+
+interface StoreDispatcher {
```

Comment:
@rahulsainani @MatthewTighe this is the abstraction that I've added here which allows us to pull out the exception handler and solves for one of the 'Unresolved questions' around the `StoreThreadFactory`.

Note how `StoreDispatcher` has a `DefaultStoreDispatcher` that will hold all the existing code in the current Store, and the `UiStoreDispatcher` will be a simplified version that holds references to the main thread.

Reply:
We can continue to discuss this after the RFC lands as it's implementation. 

In this interface, we could only expose`scope` and the rest is implemented in `DefaultStoreDispatcher` and `UiStoreDispatcher`